datadog-ffe-ffi/src/assignment.rs

```diff
@@ -13,48 +14,59 @@ use crate::Handle;
+    // Pin the inner result to ensure the memory address remains stable,
+    // allowing us to safely store borrowed pointers to its contents.
+    inner: Pin<Box<Result<Assignment, EvaluationError>>>,
     // memoizing some fields, so we can hand off references to them:
     error_message: Option<String>,
     extra_logging: Vec<KeyValue<BorrowedStr, BorrowedStr>>,
     flag_metadata: Vec<KeyValue<BorrowedStr, BorrowedStr>>,
 }
 impl ResolutionDetails {
     fn new(value: Result<Assignment, EvaluationError>) -> ResolutionDetails {
-        let error_message = value.as_ref().err().map(|err| err.to_string());
+        // Pin the value to the heap to ensure its memory address remains stable
+        let inner = Box::pin(value);
 
-        let extra_logging = value
+        let error_message = (*inner).as_ref().err().map(|err| err.to_string());
+
+        let extra_logging = (*inner)
             .as_ref()
             .iter()
             .flat_map(|it| it.extra_logging.iter())
             .map(|(k, v)| {
                 KeyValue {
-                    // SAFETY: the borrow is valid as long as string allocation is
-                    // alive. ResolutionDetails will get moved into heap but this does not
-                    // invalidate the string.
+                    // SAFETY: The returned BorrowedStr does not outlive the source string k. The
+                    // source string k lives in the pinned Assignment inside `inner`, which is owned
+                    // by ResolutionDetails and will not move (guaranteed by Pin) or be modified (we
+                    // only hold shared references).
                     key: unsafe { BorrowedStr::borrow_from_str(k.as_str()) },
-                    // SAFETY: the borrow is valid as long as string allocation is
-                    // alive. ResolutionDetails will get moved into heap but this does not
-                    // innvalidate the string.
+                    // SAFETY: The returned BorrowedStr does not outlive the source string v. The
+                    // source string v lives in the pinned Assignment inside `inner`, which is owned
+                    // by ResolutionDetails and will not move (guaranteed by Pin) or be modified (we
+                    // only hold shared references).
                     value: unsafe { BorrowedStr::borrow_from_str(v.as_str()) },
                 }
             })
             .collect();
 
-        let flag_metadata = match value.as_ref() {
+        let flag_metadata = match (*inner).as_ref() {
             Ok(a) => {
                 vec![KeyValue {
-                    // SAFETY: borrowing from static is safe as it lives long enough.
+                    // SAFETY: The returned BorrowedStr does not outlive the source string. The
+                    // string "allocation_key" is a static string literal with 'static lifetime.
                     key: unsafe { BorrowedStr::borrow_from_str("allocation_key") },
```

Comment:
Optional and unrelated to this memory fix, but while you're here and releasing a patch, I'd suggest checking if this is camel case

```suggestion
                    key: unsafe { BorrowedStr::borrow_from_str("allocationKey") },
```

to match the test cases, e.g.

https://github.com/DataDog/libdatadog/blob/7e0063f65e05ed88099bbf71e5c1f276c0918e76/datadog-ffe/tests/data/tests/test-case-flag-with-empty-string.json#L29-L30

I see that there's a `#[serde(rename_all = "camelCase")]` on all the models that does this conversion automatically, but I don't think similar conversions happen automatically in the FFI

Reply:
Oh never mind, I see you're removing these keys here! https://github.com/DataDog/libdatadog/pull/1369